gogen/codegen.go

```diff
@@ -0,0 +1,429 @@
+	}
+
+	usedEnumeratedTypes := map[string]bool{}
+	// generatedUnions stores a map, keyed by the output name for a union,
+	// that has already been output in the generated code. This ensures that
+	// where two entities re-use a union that has already been created (e.g.,
+	// a leafref to a union) then it is output only once in the generated code.
+	generatedUnions := map[string]bool{}
+	enumTypeMap := map[string][]string{}
+	structSnippets := []GoStructCodeSnippet{}
+
+	isBuiltInType := func(fType string) bool {
+		_, ok := validGoBuiltinTypes[fType]
+		return ok
+	}
+
+	// Range through the directories to find the enumerated and union types that we
+	// need. We have to do this without writing the code out, since we require some
+	// knowledge of these types to do code generation along with the values.
+	for _, directoryPath := range ir.OrderedDirectoryPathsByName() {
+		dir := ir.Directories[directoryPath]
+
+		// Generate structs.
+		if errs := checkForBinaryKeys(dir); len(errs) != 0 {
+			codegenErr = util.AppendErrs(codegenErr, errs)
+			continue
+		}
+		structOut, errs := writeGoStruct(dir, ir.Directories, generatedUnions, opts.TransformationOptions.IgnoreShadowSchemaPaths, cg.GoOptions, cg.Config.GenerateJSONSchema)
+		if errs != nil {
+			codegenErr = util.AppendErrs(codegenErr, errs)
+			continue
+		}
+		structSnippets = append(structSnippets, structOut)
+
+		// Record down all the enum types we encounter in each field.
+
+		// definedUnionTypes keeps track of which unions we have
+		// already processed to avoid processing the same one twice.
+		definedUnionTypes := map[string]bool{}
+		for _, fn := range dir.OrderedFieldNames() {
+			field := dir.Fields[fn]
+
+			// Strip the module name from the path.
+			schemaPath := util.SlicePathToString(append([]string{""}, strings.Split(field.YANGDetails.Path, "/")[2:]...))
+			switch {
+			case field.LangType == nil:
+				// This is a directory, so we continue.
+				continue
+			case field.LangType.IsEnumeratedValue:
+				usedEnumeratedTypes[field.LangType.NativeType] = true
+				enumTypeMap[schemaPath] = []string{field.LangType.NativeType}
+			case len(field.LangType.UnionTypes) > 1:
+				if definedUnionTypes[field.LangType.NativeType] {
+					continue
+				}
+				definedUnionTypes[field.LangType.NativeType] = true
+
+				for ut := range field.LangType.UnionTypes {
+					if !isBuiltInType(ut) {
+						// non-builtin union types are always enumerated types.
+						usedEnumeratedTypes[ut] = true
+						if enumTypeMap[schemaPath] == nil {
+							enumTypeMap[schemaPath] = []string{}
+						}
+						enumTypeMap[schemaPath] = append(enumTypeMap[schemaPath], ut)
+					}
+				}
+				// Sort the enumerated types into schema order.
+				sort.Slice(enumTypeMap[schemaPath], func(i, j int) bool {
+					return field.LangType.UnionTypes[enumTypeMap[schemaPath][i]] < field.LangType.UnionTypes[enumTypeMap[schemaPath][j]]
+				})
+			}
+		}
+	}
+
+	processedEnums, err := genGoEnumeratedTypes(ir.Enums)
```

Comment:
Here we could make this `genEnumeratedTypes`, but these internal changes seem a bit less important than the public API suggestions.

Reply:
Same as above, I thought specifying that this is Go generated code, while redundant, is slightly clearer to a new reader of the codebase, which I would assume to be the vast majority of the audience.

Reply:
No strong preference here :-)